components/nimbus/android/src/main/java/org/mozilla/experiments/nimbus/FeatureVariables.kt

```diff
@@ -106,20 +181,126 @@ interface Variables {
      */
     fun getDrawable(key: String): Drawable? = null
 
+    /**
+     * Uses `getStringList(key: String)` to get a list of strings, then coerces the
+     * strings in the list into Drawables. Values that cannot be coerced are omitted.
+     */
+    fun getDrawableList(key: String): List<Drawable>? = null
+
+    /**
+     * Uses `getStringList(key: String)` to get a list of strings, then coerces the
+     * values into Drawables. Values that cannot be coerced are omitted.
+     */
+    fun getDrawableMap(key: String): Map<String, Drawable>? = null
+
     // Get a child configuration object.
+
+    /**
+     * Gets a nested `JSONObject` value for this key, and creates a new `Variables` object. If
+     * the value at the key is not a JSONObject, then return `null`.
+     */
     fun getVariables(key: String): Variables? = null
+
+    /**
+     * Gets a list value for this key, and transforms all `JSONObject`s in the list into `Variables`.
+     *
+     * If the value isn't a list, then returns `null`. Items in the list that are not `JSONObject`s
+     * are omitted from the final list.
+     */
+    fun getVariablesList(key: String): List<Variables>? = null
+
+    /**
+     * Gets a map value for this key, and transforms all `JSONObject`s that are values into `Variables`.
+     *
+     * If the value isn't a `JSONObject`, then returns `null`. Values in the map that are not `JSONObject`s
+     * are omitted from the final map.
+     */
+    fun getVariablesMap(key: String): Map<String, Variables>? = null
+
     // This may be important when transforming in to a code generated object.
-    fun <T> getVariables(key: String, transform: (Variables) -> T) = getVariables(key)?.let(transform)
+    fun <T> getVariables(key: String, transform: (Variables) -> T?) = getVariables(key)?.let(transform)
+
+    fun <T> getVariablesList(key: String, transform: (Variables) -> T?): List<T>? =
+        getVariablesList(key)?.mapNotNull(transform)
+
+    fun <T> getVariablesMap(key: String, transform: (Variables) -> T?): Map<String, T>? =
+        getVariablesMap(key)?.mapValues(transform)
 }
 
+inline fun <reified T : Enum<T>> String.asEnum(): T? = try {
+    enumValueOf<T>(this)
+} catch (e: IllegalArgumentException) {
+    null
+}
+
+/**
+ * Uses `getString(key: String)` to find a string value for the given key, and coerce it in to
+ * the `Enum<T>`. If the value doesn't correspond to a variant of the type T, then `null` is
+ * returned.
+ */
+inline fun <reified T : Enum<T>> Variables.getEnum(key: String): T? =
+    getString(key)?.asEnum<T>()
+
+/**
+ * Uses `getStringList(key: String)` to find a value that is a list of strings for the given key,
+ * and coerce each item into an `Enum<T>`.
+ * If the value doesn't correspond to a variant of the list, then `null` is
+ * returned.
+ * Items of the list that are not underlying strings, or cannot be coerced into variants,
```

Comment:
I'm curious in terms of minimizing damage when things go wrong and the variables include strings that cannot be coerced into variants... is it better if we return an error if we encounter a variable like that? or at least log or report it somewhere to flag that an experiment is faulty. It might not be worth the effort if we think our experiment review process will prevent this and hopefully with the manifest work that will never happen

Reply:
Use of Nimbus involves lots of fallbacks: the app must be able to handle when the client is not involved in an experiment, or in a control branch.

Additionally, Nimbus should never be the cause of a crash: faulty configuration can be fed to the client (via experimenter) at any time, and the app should never crash because of it (that a particular branch causes a crash is another matter, however).

So, Nimbus is one of the few places, I think, where it's okay to failing fast is the wrong thing to do.